### PR TITLE
[cm]  Conv1dGeneral implementation

### PR DESCRIPTION
**Pull request by christhetree**
This PR adds a Conv1D implementation that supports causal convolutions and cached convolutions. The convolution can be toggled to be cached or not at any point in time via the method `set_cached()`. Behaves identically to torch.nn.Conv1d when not in cached mode and causal is False. When causal is True, the convolution is padded on the left side only. When cached and not causal, the convolution delays the output by `get_delay_samples()` samples. It's also TorchScript compatible.

This PR also includes tests to ensure the Conv1d is behaving as expected for all padding types and in cached and / or causal modes of operation. The only aspect that may not yet be compatible are strides greater than 1, this can be added in a future PR.

The code is roughly based off @francescopapaleo 's Conv1dSwitching code and @hyakuchiki 's streaming convolution code.
A future PR will refactor the TCN library in the SDK and will swap out all conv1d's to use this implementation instead.

### CODE REVIEW

File: neutone_sdk/conv.py [hunk 1]
```diff
@@ -0,0 +1,365 @@
...
+                (batch_size, self.n_ch, self.padding_l))
+            self.pad_r_buf = self.pad_r_buf.new_zeros(
+                (batch_size, self.n_ch, self.padding_r))
+        else:
+            self.pad_l_buf.zero_()
+            self.pad_r_buf.zero_()
+
+    def prepare_for_inference(self) -> None:
+        self.debug_mode = False
+        self.reset()
```
Comment (bogdanteleaga) [inline neutone_sdk/conv.py:61]: Below you have a comment saying torchscript needs to pass in the batch size here, but this one is fine?

File: neutone_sdk/conv.py [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+        self.cached = cached
+        self.debug_mode = debug_mode
+
+        padded_kernel_size = (kernel_size - 1) * dilation
+        padding_l, padding_r = self._calc_padding(
+            kernel_size, stride, padding, dilation, causal)
+        # The left padding required for cached mode is the maximum of the kernel size
+        # and the specified left padding for the convolution.
+        padding_l_cached = max(padded_kernel_size, padding_l)
+        # if padding_r == 0:
```
Comment (bogdanteleaga) [inline neutone_sdk/conv.py:146]: I think you forgot about these two lines here.

File: neutone_sdk/conv.py [hunk 3]
```diff
@@ -0,0 +1,365 @@
...
+    def prepare_for_inference(self) -> None:
+        """
+        Prepares the convolution for inference by disabling debug mode and ensuring the
+        convolution is in cached mode.
+        """
+        if not self.is_cached():
+            log.info(f"Converting Conv1dGeneral to cached in prepare_for_inference()")
+            self.set_cached(True)
+        self.debug_mode = False
+        self.padding_cached.prepare_for_inference()
```
Comment (bogdanteleaga) [inline neutone_sdk/conv.py:272]: Might be unavoidable, but this should also reset it and inside this function you don't pass in the batch size anymore unlike what happens with the reset in this class.

File: neutone_sdk/conv.py [hunk 4]
```diff
@@ -0,0 +1,365 @@
...
+            batch_size: If provided, the cached padding will be resized to match the new
+                        batch size.
+        """
+        self.padding_cached.reset(batch_size)
+
+    @tr.jit.export
+    def get_delay_samples(self) -> int:
+        """
+        Returns the number of samples that the convolution delays the output by. This
+        should always be 0 when the convolution is causal. This is ill-defined when not
```
Comment (bogdanteleaga) [inline neutone_sdk/conv.py:257]: Could you add a warning for the ill-defined case?

**Comment by bogdanteleaga** [comment]
LGTM, left a few small comments. Thank you for adding all the tests as well!

File: neutone_sdk/conv.py (hunk 1 above)
Comment (christhetree) [reply neutone_sdk/conv.py:61]: Yeah `prepare_for_inference` isn't converted to torchscript so it's fine to omit the optional argument. I'll add a comment to clarify this.
